### PR TITLE
feat: Add support for mTLS to API server

### PR DESCRIPTION
**Pull request by kalfonso**
This PR adds support for mTLS traffic to the API server.

- [X] Integration test in `cli_test.go`
- [X] Tested local deployment using Docker Compose: `make start`, `make connect`
- [ ] Tested against `pranadb-exemplar-pranadb` cluster in `staging`

https://github.com/squareup/pranadb-internal/issues/36

### CODE REVIEW

File: api/server.go [hunk 1]
```diff
@@ -58,22 +63,40 @@ func (s *Server) Start() error {
...
+
 func (s *Server) startServer(list net.Listener) {
 	err := s.gsrv.Serve(list) //nolint:ifshort
-	s.lock.Lock()
-	defer s.lock.Unlock()
-	s.started = false
 	if err != nil {
 		log.Errorf("grpc server listen failed: %v", err)
 	}
+	s.lock.Lock()
```
Comment (purplefox) [inline api/server.go:97]: why the change here?

File: api/server.go [hunk 2]
```diff
@@ -58,22 +63,40 @@ func (s *Server) Start() error {
...
 	reflection.Register(s.gsrv)
 	service.RegisterPranaDBServiceServer(s.gsrv, s)
 	s.started = true
 	go s.startServer(list)
 	return nil
 }
 
+func (s *Server) getTLSOpts() ([]grpc.ServerOption, error) {
+	if len(s.tlsConfig.CACert.Contents) == 0 {
+		return []grpc.ServerOption{}, nil
```
Comment (purplefox) [inline api/server.go:80]: Why do we turn off TLS if no CA cert is provided? For TLS we should only need, at minimum a server cert and a server key?

File: client/client.go [hunk 3]
```diff
@@ -365,6 +374,17 @@ func (c *Client) RegisterProtobufs(ctx context.Context, in *service.RegisterProt
 	return errors.WithStack(err)
 }
 
+func (c *Client) getTLSDialOptions() (grpc.DialOption, error) {
+	if len(c.tlsConfig.CACert.Contents) == 0 {
+		return grpc.WithInsecure(), nil
```
Comment (purplefox) [inline client/client.go:379]: Defaulting to no TLS if no CA cert?

File: conf/tls/tls_config.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+	"github.com/squareup/pranadb/errors"
+
+	"github.com/alecthomas/kong"
+)
+
+// TLSConfig can be used to build a TLS configuration.
+type TLSConfig struct {
+	// nolint: golint
+	CACert     NamedFileContent `help:"PEM-encoded CA certificates to use for ${tls_target}." placeholder:"FILE"`
```
Comment (purplefox) [inline conf/tls/tls_config.go:16]: Is this mandatory?

File: conf/tls/tls_config.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	Key        NamedFileContent `help:"Private PEM-encoded key to use for ${tls_target}. Only use if unsetting --key-pair." placeholder:"FILE"`
```
Comment (purplefox) [inline conf/tls/tls_config.go:17]: It's unclear whether these are actual keys or paths to key files on the filesystem. It would be very helpful if there was a usage example. Could you create an example config file in the cfg directory to show usage?

File: cmd/prana/main.go [hunk 6]
```diff
@@ -15,6 +16,7 @@ var CLI struct {
 	Shell       commands.ShellCommand       `cmd:"" help:"Start a SQL shell for Prana"`
 	UploadProto commands.UploadProtoCommand `cmd:"" help:"Upload a protobuf file descriptor set that can be used by Prana to decode sources"`
 	Addr        string                      `help:"Address of PranaDB server to connect to." default:"127.0.0.1:6584"`
+	TLSConfig   tls.TLSConfig               `help:"TLS client configuration" embed:"" prefix:""`
```
Comment (purplefox) [inline cmd/prana/main.go:19]: Can you show a usage example of how the CLI would be configured at the command line with TLS? It's not clear to me by looking at this.

File: conf/tls/tls_config.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+	"NoClientCert":               tls.NoClientCert,
+	"RequestClientCert":          tls.RequestClientCert,
+	"RequireAnyClientCert":       tls.RequireAnyClientCert,
+	"VerifyClientCertIfGiven":    tls.VerifyClientCertIfGiven,
+	"RequireAndVerifyClientCert": tls.RequireAndVerifyClientCert,
+	"":                           tls.RequireAndVerifyClientCert, // The default value if nothing is provided by the configuration
+}
+
+// NamedFileContent is a custom type that is similar to kong.NamedFileContentFlag but includes an error object.
+type NamedFileContent struct {
```
Comment (purplefox) [inline conf/tls/tls_config.go:32]: data in config structs should be things like strings and int that directly reflect the contents of the actual config file. They shouldn't contain fields used in processing or for other purposes, e.g. Contents and Error

File: conf/tls/tls_config.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		f.Contents = data
+	}
+
+	f.Filename = filename
+	return nil
+}
+
+// validateTLSFiles returns the first error it finds from the NamedFileContent fields. The errors are
+// populated by kong during the decode step, and are evaluated during validation.
+func (t *TLSConfig) validateTLSFiles() error {
```
Comment (purplefox) [inline conf/tls/tls_config.go:68]: All config validation should occur in conf.Config.Validate()

File: conf/tls/tls_config.go [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+	if t.Cert.Error != nil {
+		return errors.Errorf("failed to open %q: %s", t.Cert.Filename, t.Cert.Error)
+	}
+	return nil
+}
+
+// BuildTLSConfig loads CA certs and public/private key pair given options.
+//
+// Will return (nil, nil) if both cert and keyare not specified.
+func BuildTLSConfig(config TLSConfig) (*tls.Config, error) {
```
Comment (purplefox) [inline conf/tls/tls_config.go:86]: I don't think this belongs here - it seems to be converting the TLS config struct into some other specific tls config struct required by some other library.

File: conf/tls/tls_config.go [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+		return nil, errors.Errorf("invalid tls client auth setting '%s'", config.ClientAuth)
+	}
+	tlsConfig.ClientAuth = clientAuth
+	tlsConfig.BuildNameToCertificate()
+
+	return tlsConfig, nil
+}
+
+// IsConfigured returns true if TLS is configured. ie. CA cert and keys are provided.
+func (t *TLSConfig) IsConfigured() bool {
```
Comment (purplefox) [inline conf/tls/tls_config.go:123]: Not sure what config.IsConfigured() means. Isn't all config, by definition, configured?
Perhaps it should be called hasTLS()

Comment (purplefox) [reply conf/tls/tls_config.go:123]: Also, I don't think it belongs here as only used when converting to another type of tls config required for grpc

File: api/server.go (hunk 1 above)
Comment (kalfonso) [reply api/server.go:98]: unintentional. Fixed

File: api/server.go (hunk 2 above)
Comment (kalfonso) [reply api/server.go:81]: Fixed this. Now it checks the API TLS config is not empty to decide to turn TLS on/off. 

It requires CA cert to verify certificate signature.

File: client/client.go (hunk 3 above)
Comment (kalfonso) [reply client/client.go:379]: Fixed this. Now it checks the API TLS config is not empty to decide to turn TLS on/off.

File: conf/tls/tls_config.go (hunk 4 above)
Comment (kalfonso) [reply conf/tls/tls_config.go]: Yes, we [add the CA cert to the root CAs](https://go.dev/play/p/Tk9CR4BUyU). This is so that the clients can validate the server certificate.

File: conf/tls/tls_config.go (hunk 5 above)
Comment (kalfonso) [reply conf/tls/tls_config.go]: Done. Added it to `local-deployment/conf/pranadb-conf`. This is the deployment using Docker compose.

The API TLS can be configured inside the PranaDB config HCL or overridden via CLI arguments using `--api-tls-ca-cert`, `--api-tls-key`. The PranaDB conf struct uses Kong's struct `embed` tag with a prefix `api-tls-`.

File: cmd/prana/main.go (hunk 6 above)
Comment (kalfonso) [reply cmd/prana/main.go:19]: Explained above. The API TLS can use CLI arguments prefixed with `api-tls-`: `--api-tls-ca-cert`, `--api-tls-key`, etc. 
I'm planning on using different prefix for Raft TLS: `raft-tls-ca-cert`, `raft-tls-key`, etc.

File: conf/tls/tls_config.go (hunk 7 above)
Comment (kalfonso) [reply conf/tls/tls_config.go:32]: This uses a variation of [Kong CLI's NamedFileContentFlag](https://github.com/alecthomas/kong/blob/a05a0c20ba2d16bce52fbb6e5f59b25a08553880/mapper.go#L776). It implement's [Kong's `MapperValue` interface](https://pkg.go.dev/github.com/alecthomas/kong#MapperValue) to resolve the file's content and return an error if it fails to read the files.

File: conf/tls/tls_config.go (hunk 8 above)
Comment (kalfonso) [reply conf/tls/tls_config.go]: Fixed

File: conf/tls/tls_config.go (hunk 9 above)
Comment (kalfonso) [reply conf/tls/tls_config.go:86]: This converts the TLS config in the PranaDB configuration (cert files) to [Golang's crypto/tls Config](https://pkg.go.dev/crypto/tls#Config). This is then passed to the grpc TLS credentials configuration:

```
tlsConfig, err := tls.BuildTLSConfig(s.tlsConfig)
	if err != nil {
		return nil, errors.WithStack(err)
	}
	opts := []grpc.ServerOption{
		grpc.Creds(credentials.NewTLS(tlsConfig)),
	}
```

File: conf/tls/tls_config.go (hunk 10 above)
Comment (kalfonso) [reply conf/tls/tls_config.go]: Removed

File: cmd/prana/main.go (hunk 6 above)
Comment (purplefox) [reply cmd/prana/main.go:19]: raft tls?
I don't think there is a requirement to encrypt raft tls at this point

File: conf/tls/tls_config.go [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+	"github.com/squareup/pranadb/errors"
+
+	"github.com/alecthomas/kong"
+)
+
+// TLSConfig can be used to build a TLS configuration.
+type TLSConfig struct {
+	// nolint: golint
+	CACert     NamedFileContent `help:"PEM-encoded CA certificate."`
```
Comment (purplefox) [inline conf/tls/tls_config.go:16]: I still don't really understand why we need CACert.

When I configure a golang http server with TLS I only pass in server cert and server key. Example here: https://github.com/cashapp/pranadb/blob/http_api/api/http/server.go#L80

So I don't really understand why we need an extra cert for gRPC

File: conf/tls/tls_config.go (hunk 7 above)
Comment (purplefox) [reply conf/tls/tls_config.go:32]: We shouldn't use Kong for loading certs from paths. Not everyone will be using Kong and Prana needs to be embeddable and usable without using Kong - Kong should basically just parse args and load the prana config file and that's it.
Take a look at how I've configured TLS for the new HTTP API (WIP), it's super simple:
https://github.com/cashapp/pranadb/blob/http_api/conf/conf.go#L68
https://github.com/cashapp/pranadb/blob/http_api/conf/tls_conf.go
There seems to be a lot of unnecessary complexity in the way things are configured here. 
To read the actual cert is a one liner and can be done when required

Comment (kalfonso) [reply conf/tls/tls_config.go:32]: I can remove using the `MapperValue` interface to resolve the file contents. The entire PranaDB configuration is parsed by Kong library at the moment. The use of Kong is internal to PranaDB. The `prana` and `pranadb` main entry points only receive the `--config` and other flags defined in the configuration:

```
❯ go run ./cmd/pranadb/ --help
Usage: pranadb --config=CONFIG-FLAG

Flags:
  -h, --help                                                           Show context-sensitive help.
      --config=CONFIG-FLAG                                             Path to config file
      --log-format="text"                                              Format to write log lines in
      --log-level="info"                                               Lowest log level that will be emitted
      --log-file="-"                                                   File to direct logs to. If left blank, or '-', logs will go to stdout
      --node-id=INT
      --cluster-id=UINT-64
      --raft-addresses=RAFT-ADDRESSES,...
      --notif-listen-addresses=NOTIF-LISTEN-ADDRESSES,...
      --num-shards=INT
      --replication-factor=INT
      --data-dir=STRING
      --test-server
      --kafka-brokers=BROKER-CONFIGS
      --data-snapshot-entries=INT
      --data-compaction-overhead=INT
      --sequence-snapshot-entries=INT
      --sequence-compaction-overhead=INT
      --locks-snapshot-entries=INT
      --locks-compaction-overhead=INT
      --enable-api-server
      --api-server-listen-addresses=API-SERVER-LISTEN-ADDRESSES,...
      --api-tls-ca-cert=NAMED-FILE-CONTENT                             PEM-encoded CA certificate.
      --api-tls-key=NAMED-FILE-CONTENT                                 Private PEM-encoded key.
      --api-tls-cert=NAMED-FILE-CONTENT                                Public PEM-encoded cert.
      --api-tls-client-auth=STRING                                     Client certificate authentication mode, maps to the tls ClientAuthType
      --enable-source-stats
      --protobuf-descriptor-dir=STRING                                 Directory containing protobuf file descriptor sets that Prana should load to use for decoding Kafka messages. Filenames must end with .bin
      --enable-lifecycle-endpoint
      --life-cycle-listen-address=STRING
      --startup-endpoint-path=STRING
      --ready-endpoint-path=STRING
      --live-endpoint-path=STRING
      --metrics-bind="localhost:9102"                                  Bind address for Prometheus metrics ($METRICS_BIND).
      --enable-metrics
      --enable-failure-injector
      --screen-dragon-log-spam
      --disable-shard-placement-sanity-check
      --raft-rtt-ms=INT
      --raft-election-rtt=INT
      --raft-heartbeat-rtt=INT
      --disable-fsync
      --dd-profiler-types=STRING
      --dd-profiler-host-env-var-name=STRING
      --dd-profiler-port=INT
      --dd-profiler-service-name=STRING
      --dd-profiler-environment-name=STRING
      --dd-profiler-version-name=STRING
      --aggregation-cache-size-rows=INT
      --max-process-batch-size=INT
      --max-forward-write-batch-size=INT
```

File: conf/tls/tls_config.go (hunk 11 above)
Comment (kalfonso) [reply conf/tls/tls_config.go:16]: That is to enable mTLS (client provides its TLS cert to the server for authentication). For mTLS with strict validation of certificates the client and the server need to trust the CAs used to generate the client and server cert respectively.

The HTTP api server you linked does not support yet mTLS by looking at the configuration. [Here's an example using `net/http` that works through different level of TLS all the way to mTLS](https://venilnoronha.io/a-step-by-step-guide-to-mtls-in-go).

Comment (purplefox) [reply conf/tls/tls_config.go:16]: I've implemented TLS client auth many times, mainly in Java and the API is surprising to me here. Typically with client auth you would provide a set of client certificates - one for each client that is allowed to connect. Here only one cert is being passed in. Also the property is called "CACert" this is not very descriptive and not obvious what it contains - if the idea is it contains a client cert, then lets call it, for example, "ClientCert" for clarity.

File: conf/tls/tls_config.go (hunk 9 above)
Comment (purplefox) [reply conf/tls/tls_config.go:86]: We should move this into where it used as it's specific to the user, not the config itself.

File: conf/tls/tls_config.go (hunk 7 above)
Comment (purplefox) [reply conf/tls/tls_config.go:32]: Just to re-iterate the importance of this. Kong is just something we use in main() in order to load the simple config. main() is just one way to start the server. For example sqlTest does not use it. And if we embed Prana in other apps it will not use it either. That's why it's important we don't have any more complex logic for loading special contents of files in there. Given a conf.Config object with simple file paths it should be possible to start Prana.

**Comment by purplefox** [comment]
Hey @kalfonso - thanks for this. I'm going to merge it as is, then submit a new PR to integrate it with the TLS stuff in the new HTTP API so all the config is done consistently. I'll also extend the config in the process to allow multiple client certs to be specified as per the discussion above.
